### ISSUE TITLE
JS Translations in 2.8.*

### ISSUE THREAD

**Issue by HilariousGil** [ckan/ckan#4526]
I'm having issues with the translations of the JS texts. 

### CKAN Version if known (or site URL)
Branch -> 2.8
or
Master branch -> Release tag  2.8.1

### Please describe the expected behaviour
Javascript should be translated.

### Please describe the actual behaviour
Javascript texts are not being translated.

### Please describe the actual behaviour
Clean CKAN 2.8 install

### Extra Info
The configuration is set to use the public-bs2 folder.
I disabled all extensions so I can test it cleanly, and still does not work.
I'm using the this._('text to translate') function in the Javascript files, they are extracted to the .pot files but I can't seem to compile them either using the "paster trans js" or the automatic compilation when the server starts...
I have no issues with translating Jinja files.
I enabled Debug logging in core CKAN and I see the all the "JS translation for "XX" is up to date"... But then they don't show up.

Any suggestions on where to look for? does anybody else is experiencing this issue?

**Comment by HilariousGil** [comment]
Hi @klikstermkd, yes in the few places I'm using the this._('text to translate') inside the JS CKAN module.
But the behavior I'm experiencing happens also with the default core CKAN translations.
So I currently have all my plugins disabled so I'm sure nothing is worng with my code, and I'm only testing the normal CKAN translation, and that is not working properly.

That's why I didn't share any of my code, because it's all CKAN, for example:
[These line of code](https://github.com/ckan/ckan/blob/2.8/ckan/public/base/javascript/modules/confirm-action.js#L101) does not get translated.

In this image you can see the rest of the page (Jinja content) translated to Dutch, but all the JS components are in English.
![translation](https://user-images.githubusercontent.com/6845473/47566950-771dd680-d92d-11e8-9d38-6e71e851df2a.png)

Using only plain CKAN from branch 2.8

**Comment by klikstermkd** [comment]
@HilariousGil Ok I see what you mean. I tested this on https://beta.ckan.org which also runs on CKAN 2.8. The issue is a regression made with the migration to Flask. As a temporary solution you can do the following.

First in `plugin.py` override `/api/i18n/{lang}`:
```python
    # IRoutes
    def before_map(self, map):
        map.connect(
            '/api/i18n/{lang}',
            controller='ckanext.your_extension.controller:ApiController',
            action='i18n_js_translations'
        )
```

Then in `controller.py` define the `ApiController` and the appropriate `i18n_js_translations` action:
```python
from ckan.lib.base import BaseController
import os
from ckan.plugins import toolkit

class ApiController(BaseController):
    def i18n_js_translations(self, lang):
        ''' Patch for broken JS translations caused by Pylons to Flask
        migration. This method patches https://github.com/ckan/ckan/blob/master/ckan/views/api.py#L467
        :param lang: locale of the language
        :type lang: string
        :returns: the translated strings in a json file.
        :rtype: json
        '''

        ckan_path = os.path.join(
            os.path.dirname(__file__), '..', '..', '..', 'ckan', 'ckan'
        )
        source = os.path.abspath(os.path.join(ckan_path, 'public',
                                 'base', 'i18n', '%s.js' % lang))
        toolkit.response.headers['Content-Type'] =\
            'application/json;charset=utf-8'
        if not os.path.exists(source):
            return '{}'
        f = open(source, 'r')
        return(f)
```

Try this and test first if the CKAN core strings get translated, and later test your strings.

**Comment by HilariousGil** [comment]
Thanks a lot @klikstermkd !!
That worked like a charm. Should I close this issue? or is it better to keep it open for someone to look for a permanent solution?

**Comment by tino097** [comment]
it will be fixed in 2.8.2 with #4505

**Comment by ltf9651** [comment]
@klikstermkd   
Got a translation trouble with zh_CN
https://beta.ckan.org : the index home page, /user/ , /about/ page and so on won't be translated, and other page can be tranlated.
Why?
I found this by my own app and https://beta.ckan.org
